File: 46.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> output = new ArrayList<>();
        if (nums.length == 1) {
            List<Integer> singleElementList = new ArrayList<>();
            singleElementList.add(nums[0]);
            output.add(singleElementList);
            return output;
        }
        
        if (nums.length == 2) {
            List<Integer> list1 = new ArrayList<>();
            List<Integer> list2 = new ArrayList<>();
            list1.add(nums[0]);
            list1.add(nums[1]);
            list2.add(nums[1]);
            list2.add(nums[0]);
            output.add(list1);
            output.add(list2);
            return output;
        }
        
        for (int i = 0; i < nums.length; i++) {
            int[] remaining = new int[nums.length - 1];
            for (int j = 0, k = 0; j < nums.length; j++) {
                if (i != j) {
                    remaining[k++] = nums[j];
                }
            }
            List<List<Integer>> sublists = permute(remaining);
            for (List<Integer> sublist : sublists) {
                List<Integer> combinedList = new ArrayList<>();
                combinedList.add(nums[i]);
                combinedList.addAll(sublist);
                output.add(combinedList);
            }
        }
        return output;
    }
}

File: 46.py
from typing import List
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        if len(nums) == 1:
            return [nums]
        # base case:
        if len(nums) == 2:
            return [[nums[0],nums[1]],[nums[1],nums[0]]]
        else:
            out = []
            for i in range(len(nums)):
                print(i)
                print(nums[0:i] ,nums[i+1:len(nums)])
                temp = self.permute(nums[0:i] + nums[i+1:len(nums)])
                for thing in temp:
                    out.append([nums[i]]+thing)
        return out

s = Solution()
nums = [1,2,3]
print(s.permute(nums))